Vehicles: add maximum power planner hint

marcelGoerentz opened the pull request:
Fix https://github.com/evcc-io/evcc/issues/23509
Fix https://github.com/evcc-io/evcc/issues/14661

Make maximum power consumption configurable for vehicles

marcelGoerentz committed:
Merge master

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

marcelGoerentz committed:
Fix finding

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

marcelGoerentz committed:
Fix findings

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

andig commented:
For what purpose?

marcelGoerentz commented:
Planner will estimate correct time. For example my BMW iX3 can only charge with maximum 11kW. But with single phase it can be charged with 32A.
So the planner thinks it can charge also with 32A on all 3 phases. This setting will correct it to the correct value. So the planner knows the vehicle needs double the time.

mfuchs1984 commented:
> Planner will estimate correct time. 

Based on how I understand the impact of the proposed changes:

Is it really necessary to include the calculation of the current based on maximum vehicle power into setLimit for that purpose? Wouldn't it be enough to change effectiveMaxPower?

Look at e.g. https://github.com/evcc-io/evcc/discussions/25239, the car has a limit of 7.4 kW, regardless of 3 or 1 phase(s).  This results in a 3 phase current of 10.72 A. When using a charger without mA-controll, this results in 10 A. To actually achieve the 7.4 kW, the vehicle even needs at least 12 A signalled current.

Even usual 11 kW vehicles typically charge with about 10.4 to 10.6 kW on an 11 kW charger, without mA control, this seems to limit the sognalled charge current to 15 A, this charging with an even lower speed.

marcelGoerentz commented:
That's correct, changing effectiveMaxPower would be sufficient. The other changes are just for not seeing the gap between the offered current and the consumed current in the UI.

Regarding the consumed current you are right that there are losses, but we should offer the correct current anyway -> so for 11kW 16A at 3p and for 7,3kW 32A at 1p or 11A at 3p ( I know it would be 7,5kW). This way we can make sure that always enough current will be provided. We can leave the efficiency at 0,9 and this will make sure that the planner will start slightly before the "perfect" start time. Trying to estimate the correct efficiency will never work.

marcelGoerentz committed:
Add necessary files, not finished yet

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

marcelGoerentz committed:
Revert changes to loadpoint, so new option only has effect on Planner

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

marcelGoerentz committed:
Add translation for English

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

marcelGoerentz committed:
Fix nil access

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

andig commented:
I don‘t understand. It will dtill switch to fast charging and try to set 3p@32A, so you MUST limit to 16A. Doesn‘t this create more confusion?

marcelGoerentz commented:
@mfuchs1984 Yeah, you are right.

andig commented:
> There are a couple of vehicles that allow 1 phase charging with up to 32 A (e.g. our BMW i4), but only 16 A on 3 phases. 

...so you'd statically configure 32A in this case?

> With maxPower, evcc will know that there is a limit on 3 phases but is still able to use the full range for 1 phase.

It would still _only_ do this for the planner when _estimating_. It would still switch to 32A. Imho this is prone to a lot of confusion!

This (amongst others) is why I've asked:

> Would someone line to propose a description and help text for such a parameter if it were added?

I still don't see if this really is a vehicle setting or a planner hint or a no-go.

mfuchs1984 commented:
> It would still only do this for the planner when estimating.

Me too, I think this is how it implemented @marcelGoerentz

andig commented on core/loadpoint_effective.go:
```diff
@@ -232,5 +232,11 @@ func (lp *Loadpoint) EffectiveMaxPower() float64 {
 func (lp *Loadpoint) effectiveMaxPower() float64 {
+	if lp.vehicle != nil {
+		if maxPower, ok := lp.vehicle.OnIdentified().GetMaxPower(); ok {
+			lp.log.DEBUG.Printf("Max power from vehicle: %.2f kWh", maxPower/1000)
```
```suggestion
```

marcelGoerentz committed:
Add more specific help text and fix suggestion

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

marcelGoerentz committed:
Update help text

Signed-off-by: Marcel Goerentz <m.goerentz@t-online.de>

marcelGoerentz committed:
Fix punctuation in charging power descriptions

Added missing periods to the help descriptions for maximum vehicle charging power.

naltatis committed:
consistancy

naltatis reviewed (approved):
Nice solution. This should also fix https://github.com/evcc-io/evcc/issues/14661.

premultiply commented:
Was wäre da deiner Meinung nach der Unterschied?
Hier geht es in jedem Fall um die Summe und nicht um eine einzelne Phasenleistung. Daher sind DC und AC jederzeit gleich, oder?

marcelGoerentz commented:
Ist nicht das Maximum 50kW bei nicht kommerziell genutzten DC-Ladesäulen?
Mal davon abgesehen, dass vermutlich niemand eine DC Säule hat die die maximale Leistung des Autos bereitstellen kann.

Hofyyy commented:
Grundsätzlich stimme ich zu. Und sagen wir DC is max 50. Dann sind 11/22kw immer noch nicht 50. 

ich bin auch nur drauf gekommen weil wir mit Bidirektionalem kram in 1-5 jahren mehr DC sehen werden.

Daher die idee das hier gleich sauber zu framen. Definitiv kein muss. schadet ggf. aber auch nicht. kan mir bei review nur in dem kopf.
